IBX-143: Refactored reindexing command to be DBAL 2.13-compatible

alongosz opened the pull request:
| Question                                  | Answer
| ---------------------------------------- | ------------------
| **JIRA issue**                          | [IBX-143](https://issues.ibexa.co/browse/IBX-143)
| **Type**                                   | improvement
| **Target Ibexa version** | `v2.5`, `v3.2`, `v3.3`
| **BC breaks**                          | no

Since eZ Platform 2.5 [requires Doctrine Bundle ^1.9](https://github.com/ezsystems/ezplatform/blob/v2.5.13/composer.json#L28) which requires `doctrine/dbal:^2.5.12`, it's safe to force `doctrine/dbal:^2.13`, which introduces some changes ResultStatement-related class structure.

This PR besides fixing the issue at hand, refactors the code to be more future-proof and adds test coverage to prevent further regressions.

#### Checklist:
- [x] Provided PR description.
- [x] Tested the solution manually.
- [x] Provided automated test coverage.
- [x] Checked that target branch is set correctly (master for features, the oldest supported for bugs).
- [x] Ran PHP CS Fixer for new PHP code (use `$ composer fix-cs`).
- [x] Asked for a review (ping `@ibexa/engineering`).

Steveb-p commented on eZ/Publish/SPI/Search/Content/IndexerGateway.php:
```diff
@@ -0,0 +1,39 @@
+    /**
+     * @return \Generator a list of Content IDs for each iteration
+     */
+    public function fetchIteration(ResultStatement $stmt, int $iterationCount): Generator;
```
We could type-hint against an `Iterator` or a `Traversable` instead of a `Generator`. 

Our implementation returns a `Generator`, but we should be fine even with `iterable`?

Also, we should also document that we expect that it returns `Generator<int[]>`.

alongosz replied:
> We could type-hint against an `Iterator` or a `Traversable` instead of a `Generator`.
> 
> Our implementation returns a `Generator`, but we should be fine even with `iterable`?
> 
> Also, we should also document that we expect that it returns `Generator<int[]>`.

Implementation uses
```php
 $generator->next();
 $generator->valid();
```
so we can use `\Iterator` if you insist.

Steveb-p replied:
I definitely do not insist on this one. `Generator` comes with a few quirks of it's own, I'm more than fine with leaving it in this particular case. Especially since, as you noted, we use specific `Iterator` methods.

So ideally just add the phpdoc annotation about what specific data this `Generator` returns (`int[]`) and I'm more than happy :smile:

alongosz replied:
@Steveb-p unfortunately code analysis does not understand that:
![image](https://user-images.githubusercontent.com/7099219/113857240-91b53480-97a2-11eb-9dc1-ab4d393c4817.png)

Steveb-p replied:
One day, PHPStorm will respect those...

Surely...

You can add `@phpstan-return` that would be recognized by this particular tool, but... at this point I'd say it's irrelevant, thanks for bearing with me :smile:

Steveb-p reviewed (approved):
Only a minor suggestion, but otherwise LGTM.

adamwojs commented on eZ/Publish/Core/Search/Legacy/Content/IndexerGateway.php:
```diff
@@ -0,0 +1,87 @@
+        $this->connection = $connection;
+    }
+
+    public function getStatementContentSince(DateTime $since, bool $count = false): ResultStatement
```
```suggestion
    public function getStatementContentSince(DateTimeInterface $since, bool $count = false): ResultStatement
```

Please use DateTimeInterface instead of DateTime

alongosz replied:
Fixed f165feabd44271c587c553edc01c571796277f93

alongosz commented:
Update: I've reworked IndexerGateway to encapsulate all database operation, exposing only ready to consume generators. Added new namespace for SPI gateway integration test. Tests are re-using legacy Repository setup schema & data due to time constraints. Waiting for Travis...

adamwojs commented on eZ/Bundle/EzPublishCoreBundle/Command/ReindexCommand.php:
```diff
@@ -49,6 +46,19 @@ class ReindexCommand extends ContainerAwareCommand
 
+    /** @var \eZ\Publish\SPI\Search\Content\IndexerGateway */
+    private $gateway;
+    /** @var \eZ\Publish\SPI\Persistence\Content\Location\Handler */
+    private $locationHandler;
```
```suggestion

    /** @var \eZ\Publish\SPI\Persistence\Content\Location\Handler */
    private $locationHandler;
```

Nitpick: missing new line

ciastektk commented on eZ/Bundle/EzPublishCoreBundle/Command/ReindexCommand.php:
```diff
@@ -49,6 +46,19 @@ class ReindexCommand extends ContainerAwareCommand
 
+    /** @var \eZ\Publish\SPI\Search\Content\IndexerGateway */
+    private $gateway;
+    /** @var \eZ\Publish\SPI\Persistence\Content\Location\Handler */
```
```suggestion

    /** @var \eZ\Publish\SPI\Persistence\Content\Location\Handler */
```

alongosz replied:
Fixed via 260d2818

lserwatka commented on eZ/Publish/Core/Search/Legacy/Content/IndexerGateway.php:
```diff
@@ -0,0 +1,147 @@
+        return $query;
+    }
+
+    private function fetchIteration(ResultStatement $stmt, int $iterationCount): Generator
```
I would avoid name like $stmt and rather use full variable name $statement for readability.

alongosz replied:
> I would avoid name like $stmt and rather use full variable name $statement for readability.

Fixed via 319e763a

adamwojs commented on eZ/Publish/Core/Search/Legacy/Content/IndexerGateway.php:
```diff
@@ -0,0 +1,147 @@
+
+    private function fetchIteration(ResultStatement $stmt, int $iterationCount): Generator
+    {
+        do {
+            $contentIds = [];
+            for ($i = 0; $i < $iterationCount; ++$i) {
+                if ($contentId = $stmt->fetchOne()) {
+                    $contentIds[] = $contentId;
+                } elseif (empty($contentIds)) {
+                    return;
+                } else {
+                    break;
+                }
+            }
+
+            yield $contentIds;
+        } while (!empty($contentId));
```
I have feeling that something is wrong here but maybe it's lack of coffee...

Steveb-p replied:
Lack of coffee :smile: 

I was also bewildered by this code though, and for this reason wanted phpdoc to include the fact that the generator built here will actually return arrays of integers (`int[]`), not single integers (`int`). But PHPStorm refused to understand the `@return \Generator<int[]>` annotation :stuck_out_tongue_closed_eyes:

alongosz replied:
> I have feeling that something is wrong here but maybe it's lack of coffee...

It's a multidimensional array - a list of Content IDs for each iteration.

Nattfarinn replied:
Variable names are misleading, `$contentId` as an array of IDs, and `$contentIds` as an array of arrays of IDs bring a lot of confusion to this code.

Steveb-p replied:
@Nattfarinn actually no. `$contentId` is a single ID indeed.

To resolve the readability issue, I suggest we move the internal iteration to a separate, private method that will just build that `$contentIds` chunk?

Steveb-p reviewed (approved):
We can improve readability of the contested method generating `$contentIds`, but it's optional. I was able to read throught that generator just fine.
